examples/raytrace-parallel/src/lib.rs

```diff
@@ -41,304 +37,124 @@ impl Scene {
-                remaining: AtomicUsize::new(concurrency),
-                rgb_data: Mutex::new(vec![0; 4 * pixels]),
-            }),
-            ctx,
-        };
-
-        let data2 = data.clone();
-        let callback = Closure::wrap(Box::new(move |msg: Event| -> Result<(), JsValue> {
-            let mut slot = data2.borrow_mut();
-            if let Some(mut data) = slot.take() {
-                match data.event(&msg) {
-                    Ok(true) => {}
-                    Ok(false) => *slot = Some(data),
-                    Err(e) => {
-                        *slot = Some(data);
-                        return Err(e);
-                    }
-                }
-            }
-            Ok(())
-        }) as Box<dyn FnMut(_) -> Result<(), JsValue>>);
-
-        for worker in &pool.workers[..concurrency] {
-            let ptr_to_send = Arc::into_raw(r.shared.clone()) as u32;
-            let ptr_to_send = JsValue::from(ptr_to_send);
-            worker.post_message(&ptr_to_send)?;
-            worker.set_onmessage(Some(callback.as_ref().unchecked_ref()));
-            worker.set_onerror(Some(callback.as_ref().unchecked_ref()));
-        }
-
-        r.callback = Some(callback);
-        *data.borrow_mut() = Some(r);
+        let scene = self.inner;
+        let height = scene.height;
+        let width = scene.width;
+
+        // Allocate the pixel data which our threads will be writing into.
+        let pixels = (width * height) as usize;
+        let mut rgb_data = vec![0; 4 * pixels];
+        let base = rgb_data.as_ptr() as usize;
+        let len = rgb_data.len();
+
+        // Configure a rayon thread pool which will pull web workers from
+        // `pool`.
+        let thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(concurrency - 1)
+            .spawn(|thread| Ok(pool.run(|| thread.run()).unwrap()))
+            .unwrap();
+
+        // A number indicating at what granularity we're parallelizing pixels.
+        // A value of 1 means that each pixel is parallelized with respect to
+        // all the others, and a huge value would mean no parallelism at all!
+        let pixel_granularity_to_parallelize = 100;
+
+        // And now execute the render! The entire render happens on our worker
+        // threads so we don't lock up the main thread, so we ship off a thread
+        // which actually does the whole rayon business. When our returned
+        // future is resolved we can pull out the final version of the image.
+        let done = pool
+            .run_notify(move || {
+                thread_pool.install(|| {
+                    rgb_data
+                        .par_chunks_mut(pixel_granularity_to_parallelize * 4)
+                        .enumerate()
+                        .for_each(|(i, chunk)| {
+                            for (j, chunk) in chunk.chunks_mut(4).enumerate() {
+                                let i = (i * pixel_granularity_to_parallelize + j) as u32;
```

Comment:
I think you could accomplish the same granularity with less munging like this:

```rust
rgb_data
    .par_chunks_mut(4)
    .with_min_len(pixel_granularity_to_parallelize)
    .enumerate()
    .for_each(|(i, chunk)| { ... });
```

So `with_min_len` sets a lower bound on how far rayon may split the work for parallelism. The inner serial work will be a batch of *at least* that length of items. The `for_each` will still see the individual chunk-4 pixels you actually care about.

Setting a minimum length isn't always needed, but rayon tries to be adaptive and split further when stuff is getting stolen a lot. Sometimes this gets overaggressive, especially at the tail end of a computation when they steal/fight to complete the last bits.

Reply:
Oh nice, thanks!

TBH It only seems marginally better to do something like `with_min_len(100)` here and the runtime in a browser seems to oscillate between 1800ms and 1100ms locally but it seems to oscillate a little less using `with_min_len(100)`. I think in the end given how easy it is to tack this on to rayon I'll probably just leave this out